### ISSUE TITLE
mongodb interface

### ISSUE THREAD

**Issue by hydra-zone** [chihaya/chihaya#318]
some of our coders are working on a mongodb middleware for chihaya
https://git.decibel.rip/X/chihaya

at some point these new middlewares could be merged into this repo if the code quality is sufficient (?)

**Comment by mrd0ll4r** [comment]
Hey,
absolutely, yes! Do keep us posted on what you guys are doing :)
We're also always idling on IRC if you have any questions.

**Comment by hydra-zone** [comment]
I started yesterday and have the following two middlewares working:
- mongostatistics: writes a document for each combination of clientid and torrent including the total uploaded and downloaded amounts; keeps the amounts updated
- torrentapproval: (currently) reads a list of infohashes from a mongodb collection on startup and returns an error when an announce is sent for an infohash that's not listed

plans:
- torrentapproval: add http server for changes to the approved torrents list on runtime

**Comment by mrd0ll4r** [comment]
Hey,
I'm following your commits, looks good so far.

The `torrentapproval` middleware's map of torrents needs to be locked because you modify it. (This is different from the `clientapproval` middleware, which has a read-only map.) You could either use the new  `sync.Map` they added in 1.8 (I think?), or do something similar to what we're doing in the peer store. We're breaking up the 160-bit space of infohashes into *n* (say, 1024) pieces and lock them individually.

You have to keep in mind that chihaya can only perform well iff none of the (pre-)hooks limit concurrency. So you either have to program entirely lock-free or use a different technique to allow multiple parallel requests.

I have no experience with mongodb and cannot comment on its performance, especially because you'd want multiple requests in parallel.

One more thing about the `torrentapproval` middleware: You chose to not protect scrapes with the reason that a client would fail to announce. This is true, but:
1. Clients can scrape before they announce
2. Unknown infohashes return an empty scrape by default, see [storage/storage.go](https://github.com/chihaya/chihaya/blob/master/storage/storage.go#L64-L65). This could be confusing for clients, could also be used to create unnecessary load on the storage.

One final thing: You should set up some CI and use PRs :) [this line](https://git.decibel.rip/X/chihaya/src/master/middleware/torrentapproval/torrentapproval.go#L40) doesn't look like it was meant to look like that.

Keep it up! We've always wanted to write middleware for private trackers, but never got to it.

**Comment by hydra-zone** [comment]
I've run into a big problem: With my mongostatistics posthook enabled memory usage would skyrocket to multiple gigabytes with a 14,000 peers test instance. I tracked it down to every kind of find/update/upsert/insert to the mongodb via mgo. I tried creating a session every time an announce gets handled, copying a global session and always using a global session, but the result didn't get much better. 
This seems to be related to mgo itself: https://stackoverflow.com/a/21093403/7859063
I don't see a solution for this problem, I'll post multiple ways of writing the data so that others can look for errors in the code.

**Comment by mrd0ll4r** [comment]
FYI I wrote a middleware to calculate up/down deltas from announces, to be found here: https://github.com/mrd0ll4r/chihaya-privtrak-middleware
